test/integration/scheduler/framework_test.go

```diff
@@ -616,6 +667,198 @@ func TestUnreservePlugin(t *testing.T) {
+		},
+		PostBind: &schedulerconfig.PluginSet{
+			Enabled: []schedulerconfig.Plugin{{Name: postbindPlugin.Name()}},
+		},
+	}
+	// Set reserve and bind config for testing
+	pluginConfig := []schedulerconfig.PluginConfig{
+		{
+			Name: unreservePlugin.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: bindPlugin1.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: bindPlugin2.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: postbindPlugin.Name(),
+			Args: runtime.Unknown{},
+		},
+	}
+
+	// Create the master and the scheduler with the test plugin set.
+	context := initTestSchedulerWithOptions(t, testContext,
+		false, nil, registry, plugins, pluginConfig, false, time.Second)
+	defer cleanupTest(t, context)
+
+	cs := context.clientSet
+	// Add a few nodes.
+	_, err := createNodes(cs, "test-node", nil, 2)
+	if err != nil {
+		t.Fatalf("Cannot create nodes: %v", err)
+	}
+
+	tests := []struct {
+		bindPluginStatuses     []*framework.Status
+		expectBoundByScheduler bool   // true means this test case expecting scheduler would bind pods
+		expectBoundByPlugin    bool   // true means this test case expecting a plugin would bind pods
+		expectBindPluginName   string // expecting plugin name to bind pods
+		expectInvokeEvents     []pluginInvokeEvent
+	}{
+		// bind plugins skiped to bind the pod and scheduler bond the pod
+		{
+			bindPluginStatuses:     []*framework.Status{framework.NewStatus(framework.Skip, ""), framework.NewStatus(framework.Skip, "")},
+			expectBoundByScheduler: true,
+			expectInvokeEvents:     []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: bindPlugin2.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bindplugin2 succeeded to bind the pod
+		{
+			bindPluginStatuses:   []*framework.Status{framework.NewStatus(framework.Skip, ""), framework.NewStatus(framework.Success, "")},
+			expectBoundByPlugin:  true,
+			expectBindPluginName: bindPlugin2.Name(),
+			expectInvokeEvents:   []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: bindPlugin2.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bindplugin1 succeeded to bind the pod
+		{
+			bindPluginStatuses:   []*framework.Status{framework.NewStatus(framework.Success, ""), framework.NewStatus(framework.Success, "")},
+			expectBoundByPlugin:  true,
+			expectBindPluginName: bindPlugin1.Name(),
+			expectInvokeEvents:   []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bind plugin fails to bind the pod
+		{
+			bindPluginStatuses: []*framework.Status{framework.NewStatus(framework.Error, "failed to bind"), framework.NewStatus(framework.Success, "")},
+			expectInvokeEvents: []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: unreservePlugin.Name(), val: 1}, {pluginName: bindPlugin1.Name(), val: 2}, {pluginName: unreservePlugin.Name(), val: 2}},
```

Comment:
We still expect these plugins to be invoked once or twice. Don't these cause similar flakes?

Reply:
Yes, if the scheduler goroutine is too poor to get enough cpu cycles to schedule the pod twice within 30 seconds, then this test case will be failed. But I expect it will happen rarely, right?

Reply:
But In order to totally resolve the race condition, should I add a func like `initTestSchedulerWithOptions` without calling `scheduler.Run()` and export `scheduler.scheduleOne` func so that we can call it in tests?

Reply:
> Yes, if the scheduler goroutine is too poor to get enough cpu cycles to schedule the pod twice within 30 seconds, then this test case will be failed. But I expect it will happen rarely, right?

My worry is the reverse of this scenario. The scheduler may retry more than a couple of times before you check. So, it could check 3 times before the test gets a chance to check the value.

> But In order to totally resolve the race condition, should I add a func like initTestSchedulerWithOptions without calling scheduler.Run() and export scheduler.scheduleOne func so that we can call it in tests?

No, let's go with checking that `numBindCalls >= numUnreserveCalls and numUnreserveCalls >= 1`.

Reply:
This is the event checking parts I added in the second commit. If scheduler retries more than twice, it won't fail either, right?

```golang
		for j := range test.expectInvokeEvents {
			expectEvent := test.expectInvokeEvents[j]
			select {
			case event := <-pluginInvokeEventChan:
				if event.pluginName != expectEvent.pluginName {
					t.Errorf("test #%v: Expect invoke event %d from plugin %s instead of %s", i, j, expectEvent.pluginName, event.pluginName)
				}
				if event.val != expectEvent.val {
					t.Errorf("test #%v: Expect val of invoke event %d to be %d instead of %d", i, j, expectEvent.val, event.val)
				}
			case <-time.After(time.Second * 30):
				t.Errorf("test #%v: Waiting for invoke event %d timeout.", i, j)
			}
		}
```

Reply:
In the last test, isn't `expecptInvokeEvents[3].val` is 2? This is the number of times we expect the `unreserve` plugin to be called. If it is called more than 2 times, won't it cause `if event.val != expectEvent.val` to be triggered? Am I missing something?

Reply:
No, it won't. `expecptInvokeEvents[3].val` is the snapshot value of `numUnreserveCalled` when scheduler calls `unreserve` plugin. It is not an int pointer to `numUnreserveCalled`.

Reply:
Thanks for explaining. I was aware of the fact that the code was sending those `numUnreserveCalled` to a channel and compared them; what I missed was the fact that your code ignores additional values in the channel. So, if the `unreserve` plugins is called 5 times, you only care about the first two invocations that you have specified in `expectInvokeEvents`.

This is fine for this test that we expect the plugins to be called "at least" a certain number of times. This wouldn't work in other tests where a fixed number of invocations is expected.